pkg/controller/networkpolicy/crd_utils.go

```diff
@@ -42,18 +46,96 @@ func toAntreaServicesForCRD(npPorts []secv1alpha1.NetworkPolicyPort) ([]controlp
-		if npPort.Port != nil && npPort.Port.Type == intstr.String {
-			namedPortExists = true
+		if npPort.Port != nil {
+			namedPortExists = addSingleAntreaServiceForCRD(npPort.Protocol, npPort.Port, &antreaServices)
 		}
-		antreaService := controlplane.Service{
-			Protocol: toAntreaProtocol(npPort.Protocol),
-			Port:     npPort.Port,
+		if npPort.PortRange != nil {
+			if npPort.PortRange.Port != nil {
+				namedPortExists = addSingleAntreaServiceForCRD(npPort.Protocol, npPort.PortRange.Port, &antreaServices) || namedPortExists
+			} else if npPort.PortRange.From != nil && npPort.PortRange.To != nil {
+				if npPort.PortRange.Except != nil {
+					portRanges := rangeSplit(*npPort.PortRange.From, *npPort.PortRange.To, npPort.PortRange.Except)
+					for _, portRange := range portRanges {
+						err := addRangeAntreaServiceForCRD(npPort.Protocol, portRange, &antreaServices)
+						if err != nil {
+							klog.Error(err.Error())
+						}
+					}
+				} else {
+					err := addRangeAntreaServiceForCRD(npPort.Protocol, network_policy.PortRange{Start: *npPort.PortRange.From, End: *npPort.PortRange.To}, &antreaServices)
+					if err != nil {
+						klog.Error(err.Error())
+					}
+				}
+			}
+		}
+		if npPort.Port == nil && npPort.PortRange == nil {
+			antreaServices = append(antreaServices, controlplane.Service{Protocol: toAntreaProtocol(npPort.Protocol)})
 		}
-		antreaServices = append(antreaServices, antreaService)
 	}
 	return antreaServices, namedPortExists
 }
 
+// Add single antrea service with single port
+func addSingleAntreaServiceForCRD(protocol *v1.Protocol, port *intstr.IntOrString, antreaServices *[]controlplane.Service) bool {
+	var namedPortExists bool
+	if port.Type == intstr.String {
+		namedPortExists = true
+	}
+	antreaService := controlplane.Service{
+		Protocol: toAntreaProtocol(protocol),
+		PortMask: &controlplane.PortMask{Port: port},
+	}
+	*antreaServices = append(*antreaServices, antreaService)
+	return namedPortExists
+}
+
+// Add several antrea range service with a port range
+func addRangeAntreaServiceForCRD(protocol *v1.Protocol, portRange network_policy.PortRange, antreaServices *[]controlplane.Service) error {
+	bitRanges, err := portRange.BitwiseMatch()
+	if err != nil {
+		return fmt.Errorf("error when get bitMatch: %s", err.Error())
+	}
+	for _, bitRange := range bitRanges {
+		port := intstr.FromInt(int(bitRange.Value))
+		mask := int32(bitRange.Mask)
+		antreaService := controlplane.Service{
+			Protocol: toAntreaProtocol(protocol),
+			PortMask: &controlplane.PortMask{
+				Port: &port,
+				Mask: &mask,
+			},
+		}
+		*antreaServices = append(*antreaServices, antreaService)
+	}
+	return nil
+}
+
+// Transform a range with exception to several range
+func rangeSplit(from uint16, to uint16, excepts []uint16) []network_policy.PortRange {
+	var portRanges []network_policy.PortRange
+	sort.Slice(excepts, func(i, j int) bool { return excepts[i] < excepts[j] })
+	for _, except := range excepts {
+		if except < from {
```

Comment:
I would prefer `except < from` and `except > to` to be something checked in validation webhook rather than here